### PR TITLE
Adding initial tests for URLSearchParams.

### PR DESCRIPTION
**Pull request by mikewest**
These tests are mostly pulled from Sigbjorn Finne's excellent patch
at https://codereview.chromium.org/143313002/. I believe they cover
the core pieces of Mozilla's suite that @ms2ger pointed me to at
w3c/web-platform-tests#2322.

[Review on Reviewable](https://reviewable.io/reviews/w3c/web-platform-tests/2326)

### CODE REVIEW

**Comment by mikewest** [comment]
@annevk: It would be lovely if you could take a look at these.

File: url/urlsearchparams-append.html [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+test(function() {
+    var params = new URLSearchParams();
+    params.append('first', 1);
+    params.append('second', 2);
+    params.append('third', '');
+    params.append('first', 10);
+    assert_true(params.has('first'), 'Search params object has name "first"');
+    assert_equals(params.get('first'), '1', 'Search params object has name "first" with value "1"');
+    assert_equals(params.get('second'), '2', 'Search params object has name "second" with value "2"');
+    assert_equals(params.get('third'), '', 'Search params object has name "second" with value "2"');
```
Comment (annevk) [inline url/urlsearchparams-append.html]: copypasta of the message

**Comment by mikewest** [comment]
Firefox passes all of these tests, with two exceptions:

    params.append('a', 'b\0c');
    assert_equals(params + '', 'a=b%00c'); // Firefox returns `a=b`

and

    URLSearchParams(); // Firefox doesn't throw here, even though `new` wasn't used.

Do those seem like bugs in Firefox, or bugs in my understanding?

**Comment by annevk** [comment]
Those look like bugs in Firefox.

**Comment by annevk** [comment]
@bakulf ^^

**Comment by annevk** [comment]
LGTM otherwise.

File: url/resources/testharness-extras.js [hunk 2]
```diff
@@ -0,0 +1,25 @@
+/*
+Extending the W3C testharness.js with locally useful functionality.
+*/
+
+// 'Untyped' checking of exception messages; for when you want
+// to test the error message returned rather than the exact kind.
+function assert_throws_message(func, expected, name) {
```
Comment (Ms2ger) [inline url/resources/testharness-extras.js]: This is unused (as it should be).

**Comment by Ms2ger** [comment]
Review status: 0 of 10 files reviewed at latest revision, 3 unresolved discussions.

---

<sup>**[url/urlsearchparams.html, line 10 \[r4\]](https://reviewable.io:443/reviews/w3c/web-platform-tests/2326#-K303wdzi4BJOIWdRGu0)** ([raw file](https://github.com/w3c/web-platform-tests/blob/9ed6b85fcfa2462aa3b6544e6b269e8a3b799f6d/url/urlsearchparams.html#L10)):</sup>
`url/interfaces.html` should cover all this.


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/w3c/web-platform-tests/2326)

**Comment by mikewest** [comment]
@Ms2ger: Dropped `urlsearchparams.html`: you're entirely correct that it duplicated the work in `interfaces.html`. Thanks!

**Comment by jgraham** [comment]
Squash.

**Comment by mikewest** [comment]
Squished.

**Comment by bzbarsky** [comment]
> URLSearchParams(); // Firefox doesn't throw here, even though `new` wasn't used.

@mikewest Really?  It throws for me, and I would expect it to do so ever since https://bugzilla.mozilla.org/show_bug.cgi?id=916644 was fixed, a year and a half ago.  Unless you're doing this in the console in about:home in a release build (not nightly or devedition) or something as opposed to actually running the test?

> params.append('a', 'b\0c');
> assert_equals(params + '', 'a=b%00c'); // Firefox returns `a=b`

Filed https://bugzilla.mozilla.org/show_bug.cgi?id=1224643

**Comment by mikewest** [comment]
Sorry, @bzbarsky, I misread the test output. `URLSearchParams()` throws in Firefox, exactly as it's supposed to. `new URLSearchParams(DOMException.prototype);`, on the other hand, does not. See http://w3c-test.org/url/urlsearchparams-constructor.html for detail.

**Comment by bzbarsky** [comment]
> `new URLSearchParams(DOMException.prototype);`, on the other hand, does not.

Nor should it.  The `URLSearchParams` constructor takes a `(USVString or URLSearchParams)` argument.  When you pass in `DOMException.prototype`, we end up in http://heycam.github.io/webidl/#es-to-union with `V` being that `DOMException.prototype` object.  Steps 2-14 are skipped because their preconditions are not satisfied.  We get to step 15, _types_ includes a string type (`USVString`), so we try to convert to `USVString`.  That lands in http://heycam.github.io/webidl/#es-to-USVString which tries to do http://heycam.github.io/webidl/#es-to-DOMString which invokes `ToString` on `DOMException.prototype`.  Why should this throw, exactly?  In Chrome it does, because there's a `DOMException.prototype.toString` which throws if called with a `this` value that's not a `DOMException` instance.  But that's a Chrome bug as far as I can tell.  The spec for `DOMException` at http://heycam.github.io/webidl/#es-DOMException-prototype-object has nothing to say about such a `toString` method.

Note that Firefox used to have a similar bug.  We fixed it in https://bugzilla.mozilla.org/show_bug.cgi?id=997285 as part of aligning with the spec for `DOMException`, about a year and a half ago looks like.

**Comment by mikewest** [comment]
Thanks for the explanation. Filed https://crbug.com/556950

**Comment by bzbarsky** [comment]
Thanks, @mikewest !  Should get the test updated too...

**Pull request merged**